core/src/main/scala/kafka/controller/ElectionUtil.scala

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.controller
+
+import kafka.api.LeaderAndIsr
+import org.apache.kafka.common.TopicPartition
+
+case class ElectionResult(topicPartition: TopicPartition, leaderAndIsr: Option[LeaderAndIsr], liveReplicas: Seq[Int])
+
+object ElectionUtil {
+
+  def leaderForOffline(partition: TopicPartition,
+                       leaderIsrAndControllerEpochOpt: Option[LeaderIsrAndControllerEpoch],
+                       uncleanLeaderElectionEnabled: Boolean,
+                       controllerContext: ControllerContext): ElectionResult = {
```

Comment:
Let's make this method `private[this]` it shouldn't be call directly from the outside.

Reply:
Ok. I was kind of torn actually on whether this should be the actual api we expose since the other calls are just calling this iteratively.

Reply:
Yeah, I am not sure here either. I think in some cases we have unnecessary "batching" APIs or method signatures given that we batch things in the `ControllerChannelManager`. Maybe we can work toward that code structure in the controller. Where internally we deal with one partition, replica, topic, broker, etc at a time and perform batching at the edges (IO inputs and outputs).

Reply:
I should have said `ControllerBrokerRequestBatch` instead of `ControllerChannelManager` that is the class that does batching of RPCs to brokers.